Adding the ability for pledges to be tracked in Salesforce

Alexalex1432 opened the pull request:
A pledge can be manually created in Salesforce or automatically created from Funraise.  Once the pledge exists for a supporter, then donations from that supporter will count towards the pledge (start date & end date will constrain this).
Resolving FUN-9202

Commit:
Adding the ability for pledges to be tracked in Salesforce

A pledge can be manually created in Salesforce or automatically created from Funraise.  Once the pledge exists for a supporter, then donations from that supporter will count towards the pledge (start date & end date will constrain this).
Resolving FUN-9202

Alexalex1432 commented on src/classes/frDonation.cls:
```diff
@@ -133,18 +136,29 @@ public class frDonation extends frModel {
         
+        Boolean isPledge = Boolean.valueOf(request.get('pledge'));
+        if(!isPledge) {
+            Pledge__c pledge = frPledge.findActive(opp.fr_Donor__c);
```
I think this will need to change.  If an update is sent from Funraise for an old donation that was linked to a pledge at some point in Salesforce, this line would find that there isn't an active pledge anymore and unlink it.  So I think we'll only want to try this frPledge.findActive call if and only if there isn't already a pledge linked to the opportunity, which means we'll probably need to query for the opp

Alexalex1432 replied:
Ok now the boolean alreadyMappedToPledge handles this

Commit:
Forgot to add frPledge to package.xml and applied some whitespace fixes on the classes touched in this branch

Resolving FUN-9202

Alexalex1432 commented on src/classes/frPledge.cls:
```diff
@@ -0,0 +1,88 @@
+            Pledge_Donation_uq__c = opp.Id
+        );
+        try {
+            Database.upsert(pledge, Pledge__c.Pledge_Donation_uq__c , true);
```
There are 2 fields on the pledge to track the entity that created the pledge.  Since either a pledge donation or a subscription can create a pledge in Salesforce, they are lookup fields to an opportunity or subscription in Salesforce respectively.  However, you are not allowed to mark lookup fields as unique OR as external id fields, so I made a second text field with a `_uq` suffix that should always have the same value as the lookup field.  This 2 fields accomplish two things:
1) Since they are marked as unique, that enforces the fact that a pledge donation or a subscription can only ever create one pledge
2) Since they are marked as external ids, we can do this upsert on them, which removes the need to query if it exists or needs an update in these requests.  By performing the upsert on this field, it will update the existing pledge record if it exists or create if it's missing

There is a Pledge__c workflow that is in this branch to keep these fields in sync

Alexalex1432 commented on src/classes/frPledge.cls:
```diff
@@ -0,0 +1,88 @@
+                                         AND (End_Date__c = null OR End_Date__c <= TODAY)
+                                         AND Active__c = true
+                                         AND Pledge_Donation__c = null
+                                         ORDER BY CreatedDate ASC LIMIT 1
```
This query should also check to make sure the funraise__Percent_Complete__c formula field is less than 100, since we don't want to consider pledges that have reached their pledge amount as active

adelorme82 commented on src/objects/Pledge__c.object:
```diff
@@ -0,0 +1,202 @@
+&amp;&amp; 
+(ISBLANK(End_Date__c) || End_Date__c &lt;= TODAY())</formula>
+        <formulaTreatBlanksAs>BlankAsZero</formulaTreatBlanksAs>
+        <inlineHelpText>A checkbox formula field that denotes if the pledged amount has been met or exceeded and that the StartDate is after today and EndDate is before today</inlineHelpText>
```
Can you help me understand how the active bit works? If StartDate is after today, how does that work for a donation that comes in from a subscription?

Alexalex1432 replied:
Active is defined as:
-The start date is before today (or is empty)
-The end date is after today (or is empty)
-The pledge amount hasn't been 100% fulfilled

So if a supporter has a pledge and the start date is AFTER today (or in other words, the pledge is happening in the future) then any donations (including subscription recurrences) would not get automatically matched to that pledge since it would not be considered active. Does that answer it?

Alexalex1432 replied:
Fixed the greater than and less than switch up here

Commit:
Add functionality so an opportunitiy doesn't get unmapped if it's already mapped to a pledge (either manually or automatic)

Also fixing formula field on pledge so that it is using the correct less than & greater than comparisons for start date & end date
Resolving FUN-9202
